src/ngraph/op/embedding.cpp

```diff
@@ -0,0 +1,55 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/op/embedding.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+void op::Embedding::validate_and_infer_types()
+{
+    element::Type result_et = get_input_element_type(1);
+
+    const PartialShape& arg0_shape = get_input_partial_shape(0);
+    const PartialShape& arg1_shape = get_input_partial_shape(1);
+
+    PartialShape result_shape;
+
+    if (arg0_shape.rank().is_static() && arg1_shape.rank().is_static())
+    {
+        NODE_VALIDATION_ASSERT(this, static_cast<size_t>(arg1_shape.rank()) == 2)
+            << "weights are expected to be a matrix";
+        std::vector<Dimension> result_dims(static_cast<size_t>(arg0_shape.rank()) + 1);
+        for (size_t i = 0; i < static_cast<size_t>(arg0_shape.rank()); i++)
+        {
+            result_dims[i] = arg0_shape[i];
+        }
+        result_dims[result_dims.size() - 1] = arg1_shape[1];
```

Comment:
(In conjunction with the above change), replace `= arg1_shape[1]` with `= (arg1_shape.rank().is_dynamic() ? Dimension::dynamic() : arg1_shape[1])`.

That way, even if we we don't know arg1's rank, but we know that arg0's shape is "(d1,...,dn)" we can still "provisionally" fill the output type in with "(d1,...,dn,?)".